app/Main.hs

```diff
@@ -0,0 +1,151 @@
+
+    replaceExclamation :: Text.Text -> Text.Text -> Text.Text
+    replaceExclamation boardText spotText = replaceOne "!" spotText boardText
+
+    unfilledBoard =
+      "  0   1   2 \n\
+      \0 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \1 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \2 ! | ! | ! "
+
+nextTurn :: Player -> Player
+nextTurn Cross = Circle
+nextTurn Circle = Cross
+
+replaceOne :: Text.Text -> Text.Text -> Text.Text -> Text.Text
+replaceOne pattern substitution text = Text.concat [front, substitution, Text.drop (Text.length pattern) back]
+  where
+    (front, back) = Text.breakOn pattern text
+
+playAtSpot :: Coordinates -> Player -> Board -> Maybe Board
+playAtSpot coord spot (Board board) = case board !? coord of
+  Nothing -> Just $ Board $ insert coord spot board
+  Just _ -> Nothing
+
+unrepeat :: Eq a => [a] -> Maybe a
+unrepeat (x : xs) = guard (all (== x) xs) >> pure x
+unrepeat _ = Nothing
+
+winner :: Board -> Coordinates -> Maybe Player
+winner (Board board) (x, y) =
+  foldl1' (<|>) $ isWin <$> winningChecks
+  where
+    winningChecks
+      -- Center
+      | x == 1 && y == 1 = [row, column, diag2, diag1]
+      -- True if coordinates aren't part of diagonal
+      | (x == 1 || y == 1) && (x /= y) = [row, column]
+      -- True for diag 1
+      | x == y = [row, column, diag1]
+      -- True for diag 2
+      | otherwise = [row, column, diag2]
+
+    isWin = unrepeat <=< traverse (board !?)
+
+    diag1 = [(n, n) | n <- [0 .. 2]]
+    diag2 = [(2 - n, n) | n <- [0 .. 2]]
+    row = [(x', y) | x' <- [0 .. 2]]
+    column = [(x, y') | y' <- [0 .. 2]]
+
+getCoordsAndPlayTurn :: Board -> Player -> IO (Board, Coordinates)
+getCoordsAndPlayTurn board turn = do
+  TextIO.putStrLn $ showPlayerT turn <> " turn"
+  coord <- getInputAndValidate parseIntTuple "Please Enter space seperated coordinates: x y"
+  case playAtSpot coord turn board of
+    Nothing -> do
+      setSGR [SetColor Foreground Dull Red]
+      TextIO.putStrLn "Invalid placement"
+      setSGR [Reset]
+      getCoordsAndPlayTurn board turn
+    Just newBoard -> pure (newBoard, coord)
+
+type Parser a = (Text.Text -> Either Text.Text a)
+
+getInputAndValidate :: Parser a -> Text.Text -> IO a
+getInputAndValidate validator message = do
+  TextIO.putStrLn message
+  input <- TextIO.getLine
+  case validator input of
+    Right value -> pure value
+    Left errorMessage -> do
+      setSGR [SetColor Foreground Dull Red]
+      TextIO.putStrLn errorMessage
+      setSGR [Reset]
+      getInputAndValidate validator message
+
+parseNumber :: Parser Int
+parseNumber = inBounds <=< maybeToEither "Input has to be a number" . readMaybe . Text.unpack
+  where
+    inBounds :: Int -> Either Text.Text Int
+    inBounds x
+      | x < 3 && x >= 0 = Right x
+      | otherwise = Left "Coordinate out of bounds"
+
+parseIntTuple :: Parser Coordinates
+parseIntTuple s =
+  case Text.words s of
+    [x, y] -> liftA2 (,) (parseNumber x) (parseNumber y)
```

Comment:
I'd go with `(,) `on` parseNumber` from `Data.Function`.

Reply:
You mean 
`(liftA2 (,) \`on\` parseNumber) x y